utils/error_injection.hh

```diff
@@ -93,18 +97,86 @@ extern logging::logger errinj_logger;
  *          }, f);
  *    Expected use case: emulate custom errors like timeouts.
  *
+ * 4. inject_with_handler(name, func)
+ *    Inserts code that can wait for an event.
+ *    Requires func to be a function taking an injection_handler reference and
+ *    returning a future<>.
+ *    Expected use case: wait for an event from tests.
  */
 
 template <bool injection_enabled>
 class error_injection {
     inline static thread_local error_injection _local;
     using handler_fun = std::function<void()>;
 
+    struct received_messages_counter {
+        size_t counter{0};
+        condition_variable cv;
+    };
+
+    class injection_data;
+public:
+    /**
+     * The injection handler class is used to wait for events inside the injected code.
+     * If multiple inject_with_handler are called concurrently for the same injection_name,
+     * all of them will have separate handlers.
+     */
+    class injection_handler: public bi::list_base_hook<bi::link_mode<bi::auto_unlink>> {
+        lw_shared_ptr<received_messages_counter> _received_counter;
+        size_t _read_messages_counter{0};
+
+        explicit injection_handler(lw_shared_ptr<received_messages_counter> received_counter)
+            : _received_counter(std::move(received_counter)) {}
+
+    public:
+        template <typename Clock, typename Duration>
+        future<> wait_for_message(std::chrono::time_point<Clock, Duration> timeout) {
+            if (!_received_counter) {
+                on_internal_error(errinj_logger, "received_messages_counter is not initialized");
+            }
+
+            try {
+                co_await _received_counter->cv.wait(timeout, [&] { return _read_messages_counter < _received_counter->counter; });
+            }
+            catch (const std::exception& e) {
+                on_internal_error(errinj_logger, "Error injection wait_for_message timeout: " + std::string(e.what()));
+            }
+            ++_read_messages_counter;
+        }
+
+        friend class injection_data;
+    };
+
+private:
     struct injection_data {
         bool one_shot;
+        lw_shared_ptr<received_messages_counter> received_counter;
+        bi::list<injection_handler, bi::constant_time_size<false>> handlers;
 
         explicit injection_data(bool one_shot)
-            : one_shot{one_shot} {}
+            : one_shot(one_shot)
+            , received_counter(make_lw_shared<received_messages_counter>()) {}
+
+        void receive_message() {
+            if (received_counter) {
```

Comment:
This should be always true, right? I would turn it to `assert` (in this case I think `assert` fits better than `on_internal_error` because it's obvious that it should be true by looking a few lines above; besides, this is test-only code.)